Included a new MXBean to dynamically configure dump options

chandrams opened the pull request:
Included a new MXBean to dynamically configure dump options.  Internally it reuses the methods in com.ibm.jvm.Dump APIs to trigger dumps. 

Issue #153
Signed-off-by: Chandrakala <chandra-ms@in.ibm.com>

pshipton commented on jcl/src/com.ibm.management/share/classes/com/ibm/lang/management/DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+ * </table>
+ */
+
+public interface DiagnosticsMXBean extends PlatformManagedObject {
```
This would be better named OpenJ9DiagnosticsMXBean

pshipton commented on jcl/src/com.ibm.management/share/classes/com/ibm/lang/management/internal/DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,150 @@
+	 */
+	@Override
+	public void triggerDump(String dumpAgent) throws IllegalArgumentException {
+                switch (dumpAgent) {
```
I think there should be a security check here. Although these Dump APIs contain security checks, they are only enabled when the com.ibm.jvm.enableLegacyDumpSecurity system property is set to true, which it is not by default. Considering you can attach to a running JVM and invoke this, the proper security checks need to be in place.

pshipton commented on jcl/src/com.ibm.management/share/classes/com/ibm/lang/management/internal/DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,150 @@
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void configureDumpOptions(String dumpOptions) throws InvalidDumpOptionException, DumpConfigurationUnavailableException {
```
Should this be called setDumpOptions instead of configureDumpOptions? According to the MXBean naming conventions, a setter should be called set...
As a practical example, jconsole handles the APIs differently when they are getters, setters and boolean APIs,

pshipton commented on jcl/src/com.ibm.management/share/classes/com/ibm/lang/management/internal/DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,150 @@
+	 */
+	@Override
+	public void configureDumpOptions(String dumpOptions) throws InvalidDumpOptionException, DumpConfigurationUnavailableException {
+		System.out.println("Dump Options = " + dumpOptions);
```
This PR is marked WIP, but its been sitting for a while with no updates, not sure if you are ready for a review? The println() needs to be removed before merging.

chandrams commented:
@pshipton Sorry for the delay I was on vacation. I have a query,

Dump.setDumpOptions() did not work when I passed the below option:
"java:events=throw,filter=java/lang/NullPointerException#DumpTest.main*,range=1..1"

After investigation, I found -Xdump:dynamic has to be specified if throw/catch events are configured through Dump.setDumpOptions() when JIT is ON, whereas this option is not required when we set the above option on command line using -Xdump.

Shouldn't this be documented in Dump API or somewhere else? I could not find any documentation on this.

pshipton commented:
> Shouldn't this be documented in Dump API ?

Yes it should. I've seen this mentioned elsewhere recently as well, but can't remember where. Can you please request an update to the User Guide.

chandrams commented:
Thanks Peter. I had asked the same query here too - https://github.com/eclipse/openj9/issues/153

chandrams commented:
@pshipton - I have implemented the review comments and also fixed some synchronization issues in the test case. Please review. Thanks.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -28,13 +28,16 @@
 import com.ibm.lang.management.OpenJ9DiagnosticsMXBean;
 
 public class OpenJ9DiagnosticsMXBeanImpl implements OpenJ9DiagnosticsMXBean {
+	private static final DumpPermission DUMP_PERMISSION = new DumpPermission();
```
Checking security here is no longer required. Since my original comment, OpenJ9 has turned on the legacy security checks by default, in #979.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+	 *
+	 * @throws DumpConfigurationUnavailableException - if the configuration cannot be changed because a dump is already in progress.
+	 */
+	public void resetDumpOption() throws DumpConfigurationUnavailableException;
```
This should be resetDumpOptions with a 's', to match the jvm.Dump API being called.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+	 * This function sets options for the dump subsystem.
+	 * The dump option is passed in as a String. Use the same syntax as the -Xdump command-line option, with the 
+	 * initial -Xdump: omitted. See Using the -Xdump option as described in the section on dump agents in the 
+	 * documentation for the IBM JVM. This method may throw a DumpConfigurationUnavailableException if the dump
```
This should not refer to IBM

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+	 * This function triggers the specified dump agent.
+	 * The JVM will attempt to write the file to the specified file name. This may
+	 * include replacement tokens as documented in the section on dump agents
+	 * in the documentation for the IBM JVM.
```
Should not refer to IBM

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/PlatformMBeanProvider.java:
```diff
@@ -32,6 +32,7 @@
+import com.ibm.lang.management.OpenJ9DiagnosticsMXBean;
 
 /**
  * This class implements the service-provider interface to make IBM-specific
```
Please change IBM to OpenJ9 here.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+	 * @param dumpAgent - the dump agent to be triggered
+	 * @throws IllegalArgumentException - if the specified dump agent is invalid or unsupported by this method.
+	 * @throws java.lang.RuntimeException - if the vm does not contain RAS dump support
+	 * @throws java.lang.SecurityException - if there is a security manager and it doesn't allow the checks required to take this dump
```
please use different wording, `perform` instead of `take` would work.

pshipton replied:
or `trigger`

chandrams replied:
@pshipton I have taken this description from Dump.java, Should I change take to trigger in that file too?

pshipton replied:
yes, please change it there too

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,164 @@
+                                Dump.SnapDump();
+				break;
+                        default:
+                                throw new IllegalArgumentException("Invalid or Unsupported Dump Agent Option, cannot be triggered");
```
Please externalize this message. For an example, search for `/*[MSG` and also see openj9/jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,164 @@
+                                fileName = Dump.snapDumpToFile(fileNamePattern);
+				break;
+                        default:
+                                throw new IllegalArgumentException("Invalid or Unsupported Dump Agent Option, cannot be triggered");
```
Please externalize the message

pshipton commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,569 @@
+ *
+ */
+
+@Test(groups = { "level.sanity" })
```
I don't think this needs to be a sanity test.

chandrams replied:
I have changed it to level.extended, is this alright or is there any other group that I need to include it in?

pshipton replied:
I think its fine. The test team can comment on this.

pshipton commented:
@llxia please have someone review the tests

pshipton commented:
@chandrams are you going to fix the wording in Dump.java?

chandrams commented:
Yes, I will fix the working in Dump.java and open a separate PR.

chandrams commented:
https://github.com/eclipse/openj9/pull/1041 - Create a PR for Dump.java changes

pshipton commented:
Thanks!

chandrams commented:
@llxia Did you get a chance to review the tests?

smlambert commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,569 @@
+import org.testng.log4testng.Logger;
+import org.testng.annotations.BeforeClass;
+import org.testng.Assert;
+import org.testng.AssertJUnit;
```
You are using both Assert and AssertJUnit. You do not need both, preference to use Assert.  The AssertJUnit class is typically only used when migrating JUnit tests en masse and don't want to spend time reversing the parameters.

smlambert commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,569 @@
+
+	@AfterClass
+	public void tearDown() throws Exception {
+	}
```
If tearDown does nothing, do you need it?

smlambert commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,569 @@
+
+
+	@Test
+	public void testRemote() {
```
@llxia - we have other types of tests like this one, which start a separate process or server and expect to connect.  We should consider tagging them so we can measure if they are more flaky than other tests (grinder comparison).  They tend to be more susceptible to machine and lab glitches.

llxia replied:
Sorry for late reply. I somehow missed notification email. :(
Does it make sense to make this test in extended? Test like testSoftMxRemote is in extended too

chandrams replied:
I have changed it to extended group based on Peter's comments.

smlambert commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,569 @@
+	 * @param fileName - Dump file name
+	 * @return a boolean indicating if the file is found or not
+	 */	
+	private static boolean findFile(String dirPath, String fileName) {
```
Should this method be called findAndDeleteFile?  since it appears that is what the method does

smlambert commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,569 @@
+	 *
+	 * @param diagBean - OpenJ9DiagnosticsMXBean instance that has already been initialized
+	 */	
+	private void test_resetDumpOption(OpenJ9DiagnosticsMXBean diagBean) {
```
May want to add 's' to end of method name, to match underlying call setDumpOption**s**

chandrams commented:
@smlambert Thanks Shelley. I have made the changes as per your review comments, please verify.

chandrams commented:
Sorry missed removing tearDown, have removed it now. Thanks Shelley.

pshipton commented:
jenkins test sanity
jenkins test extended

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+ *******************************************************************************/
+
+
+package com.ibm.lang.management;
```
It just occurred to me that we shouldn't keep using the com.ibm package name for new OpenJ9 extensions, particularly when the bean is called OpenJ9DiagnosticsMXBean. Please move to a com.openj9 package instead.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+import java.lang.management.PlatformManagedObject;
+import com.ibm.jvm.DumpConfigurationUnavailableException;
+import com.ibm.jvm.InvalidDumpOptionException;
+import java.lang.RuntimeException;
```
These two imports are never used.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+ * <caption><b>Usage example for the {@link OpenJ9DiagnosticsMXBean}</b></caption>
+ * <tr> <td>
+ * <pre>
+ * {@code
```
This is generating a warning from Eclipse "javadoc: Missing closing brace for inline tag". Have you tried generating the javadoc to look for warnings and check the formatting?

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,152 @@
+import com.ibm.jvm.DumpConfigurationUnavailableException;
+import com.ibm.jvm.InvalidDumpOptionException;
+
+import com.ibm.jvm.Trace;
```
Unused import

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,152 @@
+	@Override
+	public void triggerDump(String dumpAgent) throws IllegalArgumentException {
+                switch (dumpAgent) {
+                        case "java":
```
Please add "//$NON-NLS-1$" tags to all the non-translatable strings.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,152 @@
+	 *
+	 * @return objectName representing the MXBean
+	 */
+	public ObjectName getObjectName() {
```
Needs an @Override tag to suppress the warning

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,144 @@
+	 *
+	 * @param dumpOptions - the options string to be set
+	 * @throws InvalidDumpOptionException - if the specified option cannot be set or is incorrect.
+	 * @throws DumpConfigurationUnavailableException - if the configuration cannot be changed becuase a dump is already in progress.
```
typo "becuase"

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,152 @@
+import com.ibm.jvm.Trace;
+import com.ibm.lang.management.OpenJ9DiagnosticsMXBean;
+
+public class OpenJ9DiagnosticsMXBeanImpl implements OpenJ9DiagnosticsMXBean {
```
This class should have a javadoc comment.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/PlatformMBeanProvider.java:
```diff
@@ -90,6 +91,10 @@
 			.addInterface(JvmCpuMonitorMXBean.class)
 			.register(allComponents);
 
+		ComponentBuilder.create(OpenJ9DiagnosticsMXBeanImpl.getInstance())
```
This class is only used for Java 9. com.ibm.java.lang.management.internal.ManagementUtils needs to be updated for Java 8.

pshipton commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/PlatformMBeanProvider.java:
```diff
@@ -32,9 +32,10 @@
 
 /**
- * This class implements the service-provider interface to make IBM-specific
+ * This class implements the service-provider interface to make OpenJ9-specific
```
Please fix the other 2 comments in this file which say IBM.

chandrams commented:
@pshipton Thanks for the review. I have implemented the review comments, please take a look.

javadoc - I tried with Eclipse on Windows and on command line on Linux but could not see the error. It worked fine for me.  I can send the javadoc generated through email if you would like to review.

pshipton commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,148 @@
+ * </pre></td></tr>
+ * </table>
+ */
+
```
Please remove the blank line.

pshipton commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,160 @@
+ * Implements functionality to dynamically configure dump options and to trigger supported dump agents.
+ * </p>
+ */
+
```
Please remove the blank lines.

pshipton commented:
@keithc-ca can you please take a look.

keithc-ca commented on jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties:
```diff
@@ -1158,6 +1158,7 @@ K05FE="Collection usage threshold cannot be negative."
 K05FF="Collection usage threshold cannot exceed maximum amount of memory for pool."
 K0660="Internal error while obtaining private constructor."
 K0661="Internal error while obtaining GcInfo instance."
+K0662="Invalid or Unsupported Dump Agent Option, cannot be triggered."
```
Only the first word of this sentence should be capitalized.

keithc-ca commented on jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementUtils.java:
```diff
@@ -804,6 +804,7 @@ public static void checkSupportedSingleton(Class<?> mxbeanInterface) {
 		private static final String HYPERVISOR_MXBEAN_NAME = "com.ibm.virtualization.management:type=Hypervisor"; //$NON-NLS-1$
 
 		private static final String JVM_CPU_MONITOR_MXBEAN_NAME = "com.ibm.lang.management:type=JvmCpuMonitor"; //$NON-NLS-1$
+		private static final String OPENJ9_DIAGNOSTICS_MXBEAN_NAME = "com.openj9.lang.management:type=OpenJ9Diagnostics"; //$NON-NLS-1$
```
The new bean name should not have the `com.` prefix.

keithc-ca commented on jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/PlatformMBeanProvider.java:
```diff
@@ -32,9 +32,11 @@
 import com.ibm.virtualization.management.internal.GuestOS;
 import com.ibm.virtualization.management.internal.HypervisorMXBeanImpl;
+import com.openj9.lang.management.OpenJ9DiagnosticsMXBean;
+import com.openj9.lang.management.internal.OpenJ9DiagnosticsMXBeanImpl;
```
Please remove these imports; I think they are now unused.

keithc-ca commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,148 @@
+ *   }
+ *   try {
+ *      MBeanServer mbeanServer = ManagementFactory.getPlatformMBeanServer();
+ *      if (true != mbeanServer.isRegistered(mxbeanName)) {
```
I've always consider boolean comparisons with true suspect, please don't promote it in example code; instead say
```
if (false == mbeanServer.isRegistered(mxbeanName))
```

keithc-ca commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,148 @@
+	 * A string containing the actual file name written to is returned. This may not
+	 * be the same as the requested filename for several reasons:
+	 * <ul>
+	 * <li>null or the empty string were specified, this will cause the JVM to write the
```
This should use 'was', not 'were': there *was* only one past as this case applies.

keithc-ca commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,148 @@
+	public void triggerDump(String dumpAgent) throws IllegalArgumentException;
+
+	/**
+	 * This function triggers the specified dump agent. Dump Agents supported - java, snap, system, heap.
```
Agents -> agents

keithc-ca commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,148 @@
+	 * @param dumpAgent the dump agent to be triggered
+	 * @param fileNamePattern the file name to write to, which may be null, empty or include replacement tokens
+	 * @throws InvalidDumpOptionException if the filename was invalid.
+	 * @throws IllegalArgumentException if the specified dump agent is invalid or unsupported by this method.
```
Remove the final period from the 'throws' documentation (like the ones below).

keithc-ca commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,160 @@
+ */
+
+
+public class OpenJ9DiagnosticsMXBeanImpl implements OpenJ9DiagnosticsMXBean {
```
This class should be final if possible.

keithc-ca commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,160 @@
+	@Override
+	public void triggerDump(String dumpAgent) throws IllegalArgumentException {
+                switch (dumpAgent) {
+                        case "java": //$NON-NLS-1$
```
Please don't indent case labels.

keithc-ca commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,160 @@
+	@Override
+	public String triggerDumpToFile(String dumpAgent, String fileNamePattern) throws InvalidDumpOptionException, IllegalArgumentException {
+		String fileName = null;
+                switch (dumpAgent) {
```
Please make indentation consistent.

keithc-ca commented on jcl/src/jdk.management/share/classes/com/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,160 @@
+		return fileName;
+	}
+
+	private static OpenJ9DiagnosticsMXBeanImpl instance = new OpenJ9DiagnosticsMXBeanImpl();
```
This field should be `final`.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/RemoteProcess.java:
```diff
@@ -0,0 +1,69 @@
+			} catch (UnsupportedEncodingException e) {
+				
+            		}  
+			catch (Exception e) {
```
Please format consistently: `} catch ...`.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/RemoteProcess.java:
```diff
@@ -0,0 +1,69 @@
+				new String("hello").getBytes("Unsupported");			
+
+			} catch (UnsupportedEncodingException e) {
+				
```
Please add a comment explaining why this exception is ignored.

chandrams commented:
@pshipton @keithc-ca - Implemented the review comments, please take a look. Thanks.

keithc-ca commented on jcl/src/jdk.management/share/classes/module-info.java.extra:
```diff
@@ -21,7 +21,9 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
  *******************************************************************************/
 module jdk.management {
+	requires openj9.jvm;
```
What motivates this requirement?

chandrams replied:
This mxbean reuses Dump API, hence this module requires openj9.jvm

keithc-ca replied:
We will need another strategy that doesn't necessitate this extra 'requires' clause. See https://github.com/ibmruntimes/openj9-openjdk-jdk9/pull/125.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,147 @@
+	 * handled.
+	 *
+	 * @throws DumpConfigurationUnavailableException if the configuration cannot be changed because a dump is already in progress
+         * @throws SecurityException if there is a security manager and it doesn't allow the checks required to change the dump settings
```
Please fix formatting.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,147 @@
+	 * @throws InvalidDumpOptionException if the specified option cannot be set or is incorrect
+	 * @throws DumpConfigurationUnavailableException if the configuration cannot be changed because a dump is already in progress
+	 * @throws SecurityException if there is a security manager and it doesn't allow the checks required to change the dump settings
+	 * @throws NullPointerException if options is null
```
The reference to 'options' should be to the parameter 'dumpOptions'.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,147 @@
+	public void setDumpOptions(String dumpOptions) throws InvalidDumpOptionException, DumpConfigurationUnavailableException;
+
+	/**
+	 * This function triggers the specified dump agent. Dump agents supported - java, snap, system, heap
```
Punctuation: insert 'and' and end with a period.
Ensure all text is punctuated properly.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,147 @@
+	 * @return the file name that the dump was actually written to 
+	 * @param dumpAgent the dump agent to be triggered
+	 * @param fileNamePattern the file name to write to, which may be null, empty or include replacement tokens
+	 * @throws InvalidDumpOptionException if the filename was invalid
```
This should refer to 'fileNamePattern' instead of 'filename'.
Please be consistent: 'file name' or 'filename', but not both.

chandrams replied:
Took the description from Dump.java, will make changes there as well

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,147 @@
+	 *
+	 * @return The file name of the dump that was created
+	 * @throws InvalidDumpOptionException if the dump operation fails
+	 * @throws RuntimeException if the vm does not contain RAS dump support
```
'vm' should be 'VM', or better, 'JVM'.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,156 @@
+	 */
+	@Override
+	public void triggerDump(String dumpAgent) throws IllegalArgumentException {
+                switch (dumpAgent) {
```
Please indent with leading tabs consistently.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,156 @@
+	private final static OpenJ9DiagnosticsMXBeanImpl instance = new OpenJ9DiagnosticsMXBeanImpl();
+
+	/**
+	 * Singleton accessor method. Returns an instance of {@link OpenJ9DiagnosticsMXBeanImpl}
```
Please end sentences with a period.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/RemoteProcess.java:
```diff
@@ -0,0 +1,68 @@
+
+			} catch (UnsupportedEncodingException e) {
+				// Expected exception from getBytes() hence ignored				
+            		} catch (Exception e) {
```
Please indent consistently and only with tabs.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/RemoteProcess.java:
```diff
@@ -0,0 +1,68 @@
+				System.out.println("Current dir using System:" +currentDir);
+				File file = new File(currentDir + File.separator + "catch.txt");
+				if (!file.exists()) {
+      					if (!file.createNewFile()){
```
Please fix formatting: spaces around binary operators, space before '{'.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+ * Class for testing the APIs that are provided as part of OpenJ9DiagnosticsMXBean
+ * Interface
+ *
+ */
```
Please end each sentence with a period.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+				logger.info(line);
+				list.add(line);
+			}
+		} catch (Exception e){
```
Formatting: space after 'while', before '{'.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+			mbeanServer = ManagementFactory.getPlatformMBeanServer();
+			if (true != mbeanServer.isRegistered(mxbeanName)) {
+				Assert.fail("OpenJ9DiagnosticsMXBean is not registered. Cannot Proceed.");
+			}
```
Please don't compare with true. This could just use 'Assert.assertTrue(...);'.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+
+		test_resetDumpOptions(diagBean);
+
+		test_setDumpOptions(diagBean, "local");
```
These would be better captured as 5 separate tests unless there is a reason not to do so.

chandrams replied:
These 5 methods are reused in both testLocal() and testRemote() hence haven't split them as separate tests.  If it needs to be split that way then I can add two files - TestOpenJ9DiagnosticsMXBeanLocal.java and TestOpenJ9DiagnosticsMXBeanRemote.java and make these as 5 tests.

keithc-ca replied:
I meant those should be tests that can be run independently. So instead of two methods with the `@Test` annotation, there would be ten (five local and five remote). That way we can be more confident that the tests don't interfere with each other.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+				 * name for the class OpenJ9DiagnosticsMXBean.
+				 */
+				diagBean = JMX.newMXBeanProxy(mbsc, mxbeanName, OpenJ9DiagnosticsMXBean.class);
+				if (true != mbsc.isRegistered(mxbeanName)) {
```
As above: don't compare with true.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+
+		String filename = "." + File.separator + "catch.txt";
+		File f = new File(filename);
+		if (true == f.exists()) {
```
Don't compare with true.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+				new String("hello").getBytes("Unsupported");
+			}
+		} catch (UnsupportedEncodingException e) {
+			Assert.assertTrue(true, "Set Dump Options for catch event failed as the expected exception UnsupportedEncodingException did not occur");
```
Please replace this with a comment (it is effectively dead code).

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+			File file = new File(filePath);
+			boolean exists = false;
+			while (false == exists) {
+				exists = file.exists();
```
A small delay in the loop would be reasonable.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+	private static void test_triggerDump(OpenJ9DiagnosticsMXBean diagBean) {
+		boolean found = false;
+		String[] dumpAgents = {"java", "heap", "snap", "system", "stack"};
+		String[] files = {"javacore", "heapdump", "Snap", "core"};
```
Formatting: spaces inside '{' and '}'.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+			} catch (java.lang.IllegalArgumentException e) {
+				/* stack agent is not supported by the trigger method */
+				if (agent.equals("stack")) {
+					Assert.assertTrue(true, "Did not receive the expected exception IllegalArgumentException");
```
Please replace this dead code with a comment.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+				if (agent.equals("stack")) {
+					Assert.assertTrue(true, "Did not receive the expected exception IllegalArgumentException");
+				} 
+				else {
```
Formatting: join '}' and 'else'.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+
+		try {
+			String fileName = diagBean.triggerClassicHeapDump();
+			found = findAndDeleteFile(".", new File(fileName).getName());
```
Declare 'found' here rather than above.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+			Assert.assertTrue(found, fileName + " not found");
+		} catch (InvalidDumpOptionException e){
+			Assert.fail("Unexpected InvalidDumpOptionException : " + e.getMessage());
+		} catch (Exception e){
```
Formatting: space before '{'.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+		}
+
+		List<String> newDumpOptions = new ArrayList<String>();
+		newDumpOptions = getDumpOptions();
```
Please remove the first assignment to 'newDumpOptions', just initialize with 'getDumpOptions()'.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+
+        	File[] files = dir.listFiles();
+
+        	for (File f: files) {
```
Formatting: space around ':'.

This loop breaks after the first successful delete: that doesn't seem correct. Please explain.

chandrams replied:
We check for the existence of the dump file, if found we don't loop further.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+		if (null == reader) {
+			found = false;
+		} else {
+			found = true;
```
This is redundant: if reader can only be non-null if found is already true.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+
+		logger.info(cmds);
+		try {
+			remoteServer = Runtime.getRuntime().exec(cmds);
```
I recommend you use 'ProcessBuilder' with the list of arguments to avoid problems with spaces, etc. within individual arguments.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,564 @@
+		watchdog.interrupt();
+		remoteServer.destroy();
+		try {
+			Thread.sleep(10000);
```
Must the test be delayed a full 10 seconds? Cannot we detect termination of the process without such a long wait?

chandrams commented:
@keithc-ca Thanks for the review. I have addressed the comments except for the 'requires clause' please take a look.

keithc-ca commented on test/JLM_Tests/src/org/openj9/test/java/lang/management/RemoteProcess.java:
```diff
@@ -0,0 +1,62 @@
+			} catch (UnsupportedEncodingException e) {
+				/* Expected exception from getBytes() hence ignored */
+			} catch (Exception e) {
+				e.printStackTrace();
```
What exception is anticipated here? Why should the test not fail if it is caught?

chandrams replied:
I have dynamically configured the dump options to the below event in the remote test

diagBeanRemote.setDumpOptions("java:events=catch,filter=java/io/UnsupportedEncodingException,range=1..1,file=" + dumpFilePath);

and included the below line in the RemoteProcess to get UnsupportedEncodingException:
new String("hello").getBytes("Unsupported");

Expected Exception is UnsupportedEncodingException thrown by String.getBytes()

keithc-ca replied:
Yes, UnsupportedEncodingException is caught on line 41; I was asking what is caught (that should be ignored) on line 43.

keithc-ca replied:
@chandrams Can you answer my question about that is caught on line 43?

chandrams replied:
Sorry, missed replying to this one earlier.  Will remove that catch as I don't see any other exception expected there.

keithc-ca commented:
@smlambert Can you please have another look at the updated tests to be sure you're still happy with the proposed additions?

keithc-ca commented:
It all looks good to me now, with the exception of the added 'requires' clause (which leads to the build failure).

chandrams commented:
@keithc-ca  I have removed requires openj9.jvm and made changes to use reflection.  I have removed throws from the methods as the Dump specific exceptions, I might have to change the description. Can you please take a quick look and let me know if my reflection related changes are fine.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,168 @@
+	public void setDumpOptions(String dumpOptions);
+/*[ELSE]
+	public void setDumpOptions(String dumpOptions) throws InvalidDumpOptionException, DumpConfigurationUnavailableException;
+/*[ENDIF]*/
```
This pattern of declaring checked exceptions in Java 8 and no checked exceptions in Java 9 creates a source incompatibility: code that compiles with Java 8 won't compile with Java 9 (e.g. the compiler will complain about trying to catch `InvalidDumpOptionException` that isn't thrown). To create a consistent API, those exceptions from the implementation may need to be remapped to new exception classes.

pshipton replied:
Agreed

chandrams commented:
@keithc-ca  I have made the changes as discussed, please review.

I created a openjdk9 build and executed the tests it worked fine.  Created a openjdk8 build and tested, test failed to compile with the below errors:

Test.java:44: error: package openj9.lang.management does not exist
import openj9.lang.management.*;
^
Test.java:45: error: package openj9.lang.management does not exist
import openj9.lang.management.InvalidOptionException;
                             ^
Test.java:55: error: cannot find symbol
        private static OpenJ9DiagnosticsMXBean diagBean = null;
                       ^
  symbol:   class OpenJ9DiagnosticsMXBean
  location: class Test
Test.java:56: error: cannot find symbol
        private static OpenJ9DiagnosticsMXBean diagBeanRemote = null;
                       ^
Extracted rt.jar to check for the package and classes, they were present. Tried setting the classpath explicitly to <JAVA_HOME>/jre/lib/rt.jar, compilation still failed. Copied rt.jar to another dir and specified this path in classpath, compilation worked fine.  Will try to recheck again on another machine tomorrow.

chandrams commented:
openjdk8 failure issue was due to symbol file not being updated.  On ignoring the symbol file, the code compiled. Thanks @keithc-ca for pointing me to the file where the new package needs to be updated. Will update the file and create a PR once the MXBean changes are reviewed and merged.

keithc-ca commented:
Please create the pull request for openjdk8 and add a link to it here together with information about whether the openjdk8 changes can be merged first or whether they must be merged simultaneously.

chandrams commented:
pull request for openjdk8 - https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/60

Submitted a personal build, openjdk8 build was created and all sanity tests passed.  So, I think it is fine to merge the changes required in the make. Have asked the same in the above pull request too.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,143 @@
+	 * @throws IllegalArgumentException if the specified dump agent is invalid or unsupported by this method
+	 * @throws RuntimeException if the vm does not contain RAS dump support
+	 * @throws SecurityException if there is a security manager and it doesn't allow the checks required to trigger this dump
+	 */
```
Does this also throw NPE (if dumpAgent is null)?

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,143 @@
+	 * @throws IllegalArgumentException if the specified dump agent is invalid or unsupported by this method
+	 * @throws SecurityException if there is a security manager and it doesn't allow the checks required to trigger this dump
+	 */
+	public String triggerDumpToFile(String dumpAgent, String fileNamePattern) throws IllegalArgumentException, InvalidOptionException;
```
Please document NPE (if dumpAgent is null).

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,399 @@
+		}
+/*[ELSE]
+		try {
+			Dump.resetDumpOptions();
```
Can you please move the conditional boundaries as only the body of the try statement is different (and to similar situations in other methods).

llxia commented on test/JLM_Tests/testng.xml:
```diff
@@ -56,6 +56,7 @@
 			<class name="org.openj9.test.java.lang.management.TestMisc" />
 			<class name="org.openj9.test.java.lang.management.TestManagementUtils" />
 			<class name="org.openj9.test.java.lang.management.TestLoggingMXBean" />
+			<class name="org.openj9.test.java.lang.management.TestOpenJ9DiagnosticsMXBean" />
```
Sorry for late review. I had issue with my email filtering.

TestOpenJ9DiagnosticsMXBean has annotation `level.extended` and it is added into JLM_Tests_IBMinternal in testng.xml.
In playlist.xml, we only run `sanity` for JLM_Tests_IBMinternal, which means we will not run tests in TestOpenJ9DiagnosticsMXBean.

chandrams commented:
@llxia @keithc-ca 

Implemented the review comments, please review again. Thanks

llxia commented on test/JLM_Tests/playlist.xml:
```diff
@@ -1040,6 +1040,33 @@
+			<variation>Mode100</variation>
+			<variation>Mode101</variation>
+		</variations>
+		<command>$(ADD_JVM_LIB_DIR_TO_LIBPATH) \
```
Please run test in REPORTDIR as the test may generate core file. 

```
$(MKTREE) $(REPORTDIR); \
cd $(REPORTDIR); \
```

chandrams replied:
@llxia I have made these changes, please review. Thanks

pshipton commented on jcl/src/jdk.management/share/classes/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,379 @@
+	 */
+	@Override
+	public void resetDumpOptions() throws ConfigurationUnavailableException {
+		try {
```
Although the dump APIs already check security permissions, it may be useful to have an extra layer of permission checking for the MXBean, since the MXBean can be used remotely. Please add a security check for java.lang.ManagementPermission("control") (i.e. ManagementPermissionHelper.MPCONTROL) to each of the APIs. This allows the local application to have DumpPermssion, but if ManagementPermission("control") is not granted then the MXBean APIs cannot be used.

chandrams replied:
ManagementPermissionHelper is an internal class in java.management module, not sure if there would be any restrictions in accessing internal classes using reflection from another module in future. Instead can I just use ManagementPermission("control") directly?

keithc-ca replied:
Yes, you can use `ManagementPermission("control")` directly, but you should capture that object somewhere (like ManagementPermissionHelper does) so that you can reuse it.

pshipton replied:
java.management already `exports com.ibm.java.lang.management.internal to jdk.management;` so I think ManagementPermissionHelper can be used directly.

pshipton commented:
fyi #1551, similar changes are required here

chandrams commented:
@pshipton @keithc-ca 

I have made the changes as suggested, please review. Thanks.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,145 @@
+
+/**
+ * <p>
+ * This interface provides APIs to dynamically trigger Dump agents. APIs are also available to
```
Please don't captialize 'Dump'.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,145 @@
+public interface OpenJ9DiagnosticsMXBean extends PlatformManagedObject {
+	/**
+	 * Reset the JVM dump options to the settings specified when the JVM was started removing any additional
+	 * configuration done since then. This method may throw a DumpConfigurationUnavailableException if the dump 
```
I think this should be 'ConfigurationUnavailableException' instead of 'DumpConfigurationUnavailableException'.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java:
```diff
@@ -0,0 +1,145 @@
+	 * This function sets options for the dump subsystem.
+	 * The dump option is passed in as a String. Use the same syntax as the -Xdump command-line option, with the 
+	 * initial -Xdump: omitted. See Using the -Xdump option as described in the section on dump agents in the 
+	 * documentation for the OpenJ9 JVM. This method may throw a DumpConfigurationUnavailableException if the dump
```
DumpConfigurationUnavailableException -> ConfigurationUnavailableException

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,422 @@
+			ObjectName name = new ObjectName("openj9.lang.management:type=OpenJ9Diagnostics");  //$NON-NLS-1$
+			return name;
+		} catch (MalformedObjectNameException e) {
+			return null;
```
This should probably throw an Error rather than return null.

chandrams replied:
Other MXBeans such as JVMCpuMonitor, ProcessorMXBeanImpl all have similar logic and return null on MalformedObjectNameException. Hence used the same logic here. Should I change it to throw an error?

keithc-ca replied:
I think they all should throw an Error, but I'm fine leaving that cleanup for later.

chandrams replied:
Thanks Keith. I have made the other changes as per your review comments. Please review again.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,422 @@
+/*[IF Sidecar19-SE]*/
+			try {
+				Object controlPermission = managementPermission.get(null);
+				manager.checkPermission((java.security.Permission)controlPermission);
```
Please pull the call to checkPermission() out of the `try` statement so it doesn't flow through `handleError` and is common for all Java versions.

keithc-ca commented on test/JLM_Tests/playlist.xml:
```diff
@@ -1061,6 +1061,35 @@
+		</groups>
+		<subsets>
+			<subset>SE80</subset>
+			<subset>SE90</subset>
```
This test should apply to all Java versions: can we just remove the `<subsets>` element?

smlambert replied:
No, in able to support special subsets like Panama and Valhalla, we are caught in the position of explicitly stating the subset.  If you do remove < subsets >, it will generate make targets for 5-6 supported subsets (SE80, SE90, SE100, SE110, Panama, Valhalla) and test compilation will fail, as Panama and Valhalla have special requirements for build.  (At some point we should address this in a more satisfying way, in order not to play this game of continuously updating these... perhaps Panama & Valhalla off in a separate branch of openj9? but currently explicit statement is the way)

So, in this case, you may want to add SE100 to run this test also against Java10.

keithc-ca commented on jcl/src/jdk.management/share/classes/openj9/lang/management/internal/OpenJ9DiagnosticsMXBeanImpl.java:
```diff
@@ -0,0 +1,423 @@
+			dumpConfigurationUnavailableExClass = classes[2];
+			Class<?> managementPermissionHelperClass = classes[3];
+
+			managementPermission= managementPermissionHelperClass.getField("MPCONTROL"); //$NON-NLS-1$
```
I don't think access to `ManagementPermissionHelper.MPCONTROL` needs to be via reflection. That package is exported to this module so it should be able to use the class and field directly in all Java versions. That change can wait for a separate PR though.

keithc-ca commented:
jenkins test sanity plinux
